### PR TITLE
feat(payment-stripe): coupons

### PR DESCRIPTION
**Pull request by a-sava**

### CODE REVIEW

**Comment by OlegDO** [comment]
Should be feature/ms-payment-stripe-name

image

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export default class couponEntity1691443587317 implements MigrationInterface {
+  name = 'couponEntity1691443587317';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TABLE "promo_code" ("promoCodeId" character varying(30) NOT NULL, "code" character varying NOT NULL, "couponId" character varying NOT NULL, "createdAt" TIMESTAMP NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP NOT NULL DEFAULT now(), "couponPromoCodeId" character varying(30), CONSTRAINT "PK_6aeee3e0180adf2bf461b066a82" PRIMARY KEY ("promoCodeId"))`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]:
```suggestion
      `CREATE TABLE "promo_code" ("promoCodeId" character varying(30) NOT NULL, "code" character varying NOT NULL, "couponId" character varying NOT NULL, "createdAt" TIMESTAMP NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP NOT NULL DEFAULT now(), "couponPromoCodeId" character varying(30), CONSTRAINT "promo_code(pk):promoCodeId" PRIMARY KEY ("promoCodeId"))`,
```

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TYPE "public"."coupon_duration_enum" AS ENUM('ones', 'repeating', 'forever')`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "coupon" ("couponId" character varying(8) NOT NULL, "name" character varying, "amountOff" double precision, "percentOff" integer, "duration" "public"."coupon_duration_enum" NOT NULL, "durationInMonths" integer, "maxRedemptions" integer, "createdAt" TIMESTAMP NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP NOT NULL DEFAULT now(), CONSTRAINT "PK_85a34266f36892347867b21d6ad" PRIMARY KEY ("couponId"))`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]:
```suggestion
      `CREATE TABLE "coupon" ("couponId" character varying(8) NOT NULL, "name" character varying, "amountOff" double precision, "percentOff" integer, "duration" "public"."coupon_duration_enum" NOT NULL, "durationInMonths" integer, "maxRedemptions" integer, "createdAt" TIMESTAMP NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP NOT NULL DEFAULT now(), CONSTRAINT "add correct alias" PRIMARY KEY ("couponId"))`,
```

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TABLE "coupon_products_product" ("couponId" character varying(8) NOT NULL, "productId" character varying(19) NOT NULL, CONSTRAINT "PK_0bd6a4b75fc235663fff0cdfc2b" PRIMARY KEY ("couponId", "productId"))`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]:
```suggestion
      `CREATE TABLE "coupon_products_product" ("couponId" character varying(8) NOT NULL, "productId" character varying(19) NOT NULL, CONSTRAINT "the same" PRIMARY KEY ("couponId", "productId"))`,
```

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `CREATE INDEX "IDX_006cd586575d86f4705186b869" ON "coupon_products_product" ("couponId") `,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: wrong index alias
```suggestion
      `CREATE INDEX "IDX_006cd586575d86f4705186b869" ON "coupon_products_product" ("couponId") `,
```

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `CREATE INDEX "IDX_bdd320620afd063015fa0976ae" ON "coupon_products_product" ("productId") `,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: wrong index alias

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "promo_code" ADD CONSTRAINT "FK_8db7b0ac755956ce3d0cb6266b8" FOREIGN KEY ("couponPromoCodeId") REFERENCES "promo_code"("promoCodeId") ON DELETE NO ACTION ON UPDATE NO ACTION`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: wrong fk alias

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "coupon_products_product" ADD CONSTRAINT "FK_006cd586575d86f4705186b869a" FOREIGN KEY ("couponId") REFERENCES "coupon"("couponId") ON DELETE CASCADE ON UPDATE CASCADE`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: wrong fk alias

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "coupon_products_product" ADD CONSTRAINT "FK_bdd320620afd063015fa0976ae1" FOREIGN KEY ("productId") REFERENCES "product"("productId") ON DELETE NO ACTION ON UPDATE NO ACTION`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: wrong fk alias

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE "coupon_products_product" DROP CONSTRAINT "FK_bdd320620afd063015fa0976ae1"`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: the same

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "coupon_products_product" DROP CONSTRAINT "FK_006cd586575d86f4705186b869a"`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: the same

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "promo_code" DROP CONSTRAINT "FK_8db7b0ac755956ce3d0cb6266b8"`,
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: the same

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(`DROP INDEX "public"."IDX_bdd320620afd063015fa0976ae"`);
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts]: the same

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    await queryRunner.query(`DROP INDEX "public"."IDX_006cd586575d86f4705186b869"`);
+    await queryRunner.query(`DROP TABLE "coupon_products_product"`);
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts:79]: the same

File: microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export default class couponEntity1691443587317 implements MigrationInterface {
+  name = 'couponEntity1691443587317';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
```
Comment (OlegDO) [inline microservices/payment-stripe/migrations/1691443587317-coupon-entity.ts:6]: group by query runners by action:
image

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+  @JSONSchema({
+    description: 'Field for storing id of according coupon entity created on payment service side',
+  })
+  @PrimaryColumn({ type: 'varchar', length: 8 })
+  @Length(1, 8)
+  couponId: string;
+
+  @Column({ type: 'varchar', nullable: true, default: null })
+  @IsString()
+  @IsOptional()
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  @IsUndefinable()
  @IsNullable()
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+@Entity()
+class Coupon {
+  @JSONSchema({
+    description: 'Field for storing id of according coupon entity created on payment service side',
+  })
+  @PrimaryColumn({ type: 'varchar', length: 8 })
+  @Length(1, 8)
+  couponId: string;
+
+  @Column({ type: 'varchar', nullable: true, default: null })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  @Column({ type: 'varchar', default: null })
```

Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  @Column({ type: 'uuid', nullable: true, default: null })
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+  @IsString()
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  @Length(1, 36)
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+  @IsOptional()
+  userId: string | null;
+
+  @JSONSchema({
+    description: 'Name of the coupon displayed to customers on for instance invoices or receipts',
+  })
+  @Column({ type: 'varchar', nullable: true })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: the same, @Nullable() decorator and varchar length?
```suggestion
  @Column({ type: 'varchar' })
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+  @IsString()
+  @IsOptional()
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: remove is string use @Length()
```suggestion
  @IsUndefinable()
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+  name: string | null;
+
+  @JSONSchema({
+    description: 'Amount that will be taken off the subtotal of any invoices for this customer.',
+  })
+  @Column({ type: 'float', nullable: true })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: the same decorators

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+  @IsNumber()
+  @IsOptional()
+  amountOff?: number;
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  amountOff: number | null;
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+
+  @JSONSchema({
+    description:
+      'Percent that will be taken off the subtotal of any invoices for this customer for the duration of the coupon.',
+  })
+  @Column({ type: 'int', nullable: true })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: the same decorators

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+  @IsValidate(Coupon, (coupon) => !coupon.amountOff)
+  @Min(1)
+  @Max(100)
+  @IsInt()
+  @IsOptional()
+  percentOff?: number;
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  percentOff: number | null;
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+  @IsNumber()
+  @IsOptional()
+  amountOff?: number;
+
+  @JSONSchema({
+    description:
+      'Percent that will be taken off the subtotal of any invoices for this customer for the duration of the coupon.',
+  })
+  @Column({ type: 'int', nullable: true })
+  @IsValidate(Coupon, (coupon) => !coupon.amountOff)
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: image
decompose validation
```suggestion
  @IsValidate(Coupon, (coupon) => !Coupon.isAnountOff))
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+    enum: CouponDuration,
+  })
+  @IsEnum(CouponDuration)
+  duration: CouponDuration;
+
+  @JSONSchema({
+    description: 'If duration is repeating, the number of months the coupon applies.',
+  })
+  @Column({ type: 'int', nullable: true })
+  @IsValidate(Coupon, (coupon) => coupon.duration === CouponDuration.REPEATING)
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: move out validation to private sttis method

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+    type: 'enum',
+    enum: CouponDuration,
+  })
+  @IsEnum(CouponDuration)
+  duration: CouponDuration;
+
+  @JSONSchema({
+    description: 'If duration is repeating, the number of months the coupon applies.',
+  })
+  @Column({ type: 'int', nullable: true })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  @Column({ type: 'int' })
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+  @IsValidate(Coupon, (coupon) => coupon.duration === CouponDuration.REPEATING)
+  @IsInt()
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  @Number()
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+  durationInMonths: number | null;
+
+  @JSONSchema({
+    description:
+      'Maximum number of times this coupon can be redeemed, in total, ' +
+      'across all customers, before it is no longer valid.',
+  })
+  @Column({ type: 'int', nullable: true })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]: the same @IsNumber() and is @Undefinable()

File: microservices/payment-stripe/src/entities/product.ts [hunk 29]
```diff
@@ -56,6 +59,9 @@ class Product {
 
   @OneToMany('Transaction', 'product')
   transactions: Transaction[];
+
+  @ManyToMany(() => Coupon)
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/product.ts:64]: image

File: microservices/payment-stripe/src/entities/promo-code.ts [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+
+  @IsTypeormDate()
+  @CreateDateColumn()
+  createdAt: Date;
+
+  @IsTypeormDate()
+  @UpdateDateColumn()
+  updatedAt: Date;
+
+  @ManyToOne(() => PromoCode)
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/promo-code.ts:49]: add validate nested
image

File: microservices/payment-stripe/src/entities/promo-code.ts [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+} from 'typeorm';
+import Coupon from '@entities/coupon';
+
+@JSONSchema({
+  properties: {
+    coupon: { $ref: '#/definitions/Coupon' },
+  },
+})
+@Entity()
+class PromoCode {
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/promo-code.ts:20]: update decorators

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+  IsString,
+  ValidateIf,
+} from 'class-validator';
+import { getManager } from 'typeorm';
+import CouponDuration from '@constants/coupon-duration';
+import Coupon from '@entities/coupon';
+import TCurrency from '@interfaces/currency';
+import Factory from '@services/payment-gateway/factory';
+
+class CreateCouponInput {
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/coupon/create.ts:10]: update decorators

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 33]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Create new coupon
+ */
+const create = Endpoint.custom(
+  () => ({
+    input: CreateCouponInput,
+    output: CreateCouponOutput,
+    description: 'Create new coupon',
+  }),
+  async (params) => {
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/coupon/create.ts]:
```suggestion
  async ({userId, ...}) => {
```

File: microservices/payment-stripe/src/methods/promo-code/crud.ts [hunk 34]
```diff
@@ -0,0 +1,12 @@
+import { Endpoint } from '@lomray/microservice-helpers';
+import { getRepository } from 'typeorm';
+import PromoCode from '@entities/promo-code';
+
+/**
+ * CRUD controller for Product entity
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/promo-code/crud.ts]: Product?
```suggestion
 * CRUD controller for Promo code entity
```

File: microservices/payment-stripe/src/methods/stripe/create-checkout.ts [hunk 35]
```diff
@@ -15,6 +15,10 @@ class CreateCheckoutInput {
 
   @IsString()
   cancelUrl: string;
+
+  @IsBoolean()
+  @IsOptional()
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/stripe/create-checkout.ts]:
```suggestion
  @IsUndefinable()
```

File: microservices/payment-stripe/src/services/payment-gateway/abstract.ts [hunk 36]
```diff
@@ -55,6 +57,16 @@ export interface IProductParams {
   userId: string;
 }
 
+export interface ICouponParams {
+  name?: string;
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/abstract.ts:64]: optionl params should be at the end

File: microservices/payment-stripe/src/services/payment-gateway/stripe.ts [hunk 37]
```diff
@@ -1792,6 +1805,114 @@ class Stripe extends Abstract {
       number: digits,
     };
   }
+
+  public async createCoupon({
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/stripe.ts:1812]: add jsoc

File: microservices/payment-stripe/src/services/payment-gateway/stripe.ts [hunk 38]
```diff
@@ -1792,6 +1805,114 @@ class Stripe extends Abstract {
...
+        amountOff,
+        maxRedemptions,
+        duration,
+        durationInMonths,
+      },
+      id,
+    );
+  }
+
+  private validateAndTransformCouponDiscountInput({
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/stripe.ts:1858]: add jsdoc

File: microservices/payment-stripe/src/services/payment-gateway/stripe.ts [hunk 39]
```diff
@@ -1792,6 +1805,114 @@ class Stripe extends Abstract {
...
+    percentOff,
+    amountOff,
+  }: Pick<IStripeCouponParams, 'percentOff' | 'amountOff'>): Pick<
+    StripeSdk.CouponCreateParams,
+    'amount_off' | 'percent_off'
+  > {
+    if (!amountOff && !percentOff) {
+      throw new BaseException({
+        status: 400,
+        message: 'Neither discount amount nor percent provided.',
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/stripe.ts:1868]:
```suggestion
        message: 'Neither discount amount or percent provided.',
```

File: microservices/payment-stripe/src/services/payment-gateway/stripe.ts [hunk 40]
```diff
@@ -1792,6 +1805,114 @@ class Stripe extends Abstract {
...
+
+    return {
+      // eslint-disable-next-line camelcase
+      amount_off: amountOff,
+      // eslint-disable-next-line camelcase
+      percent_off: percentOff,
+    };
+  }
+
+  private validateAndTransformCouponDurationInput({
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/stripe.ts:1890]: jsdoc

File: microservices/payment-stripe/src/services/payment-gateway/stripe.ts [hunk 41]
```diff
@@ -1792,6 +1805,114 @@ class Stripe extends Abstract {
...
+    }
+
+    return {
+      duration: duration as StripeSdk.CouponCreateParams.Duration,
+      // eslint-disable-next-line camelcase
+      duration_in_months: durationInMonths,
+    };
+  }
+
+  public async createPromoCode({ couponId, code: userCode }: IStripePromoCodeParams): Promise<{
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/stripe.ts:1916]: jsdoc

**Review by OlegDO — CHANGES_REQUESTED** [review]

File: microservices/payment-stripe/src/services/payment-gateway/stripe.ts (hunk 39 above)
Comment (a-sava) [reply microservices/payment-stripe/src/services/payment-gateway/stripe.ts:1868]: nor is corrent

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 42]
```diff
@@ -0,0 +1,138 @@
...
+  },
+})
+@Entity()
+class Coupon {
+  @JSONSchema({
+    description: 'Field for storing id of according coupon entity created on payment service side',
+  })
+  @PrimaryColumn({ type: 'varchar', length: 8 })
+  @Length(1, 8)
+  couponId: string;
```
Comment (MatthewPattell) [inline microservices/payment-stripe/src/entities/coupon.ts:31]: just `id`?

File: microservices/payment-stripe/src/entities/promo-code.ts [hunk 43]
```diff
@@ -0,0 +1,53 @@
...
+  @JSONSchema({
+    description:
+      'Field for storing id of according promo code entity created on payment service side',
+  })
+  @PrimaryColumn({ type: 'varchar', length: 30 })
+  promoCodeId: string;
```
Comment (MatthewPattell) [inline microservices/payment-stripe/src/entities/promo-code.ts:26]: just `id`?

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 44]
```diff
@@ -0,0 +1,97 @@
...
+    userId,
+    name,
+    currency,
+    amountOff,
+    percentOff,
+    duration,
+    durationInMonths,
+    maxRedemptions,
+    products,
+  }) => {
```
Comment (MatthewPattell) [inline microservices/payment-stripe/src/methods/coupon/create.ts:78]:
```suggestion
  async (params) => {
```

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 45]
```diff
@@ -0,0 +1,97 @@
...
+        userId,
+        name,
+        currency,
+        amountOff,
+        percentOff,
+        duration,
+        durationInMonths,
+        maxRedemptions,
+        products,
+      }),
```
Comment (MatthewPattell) [inline microservices/payment-stripe/src/methods/coupon/create.ts:92]:
```suggestion
      entity: await service.createCoupon(params),
```

File: microservices/payment-stripe/src/services/payment-gateway/abstract.ts [hunk 46]
```diff
@@ -225,6 +244,32 @@ abstract class Abstract {
...
+  protected async createCoupon(params: ICouponParams, couponId: string): Promise<Coupon> {
+    const products = await this.productRepository.find({
+      where: {
+        productId: In(params.products),
+      },
+    });
+
+    if (!products || products.length !== params.products.length) {
+      throw new BaseException({
+        status: 409,
```
Comment (MatthewPattell) [inline microservices/payment-stripe/src/services/payment-gateway/abstract.ts]: 409???

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 47]
```diff
@@ -0,0 +1,138 @@
...
+   * Check if amountOff exists
+   */
+  private static isAmountOffExists(entity: Coupon) {
+    return Boolean(entity.amountOff);
+  }
+
+  /**
+   * Check if duration is repeating
+   */
+  private static isDurationRepeating(entity: Coupon) {
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  private static isDurationRepeating(entity: Coupon): boolean {
```

File: microservices/payment-stripe/src/entities/coupon.ts [hunk 48]
```diff
@@ -0,0 +1,138 @@
...
+  products: Product[];
+
+  @ValidateNested()
+  @OneToMany(() => PromoCode, (promoCode) => promoCode.coupon)
+  promoCodes: PromoCode[];
+
+  /**
+   * Check if amountOff exists
+   */
+  private static isAmountOffExists(entity: Coupon) {
```
Comment (OlegDO) [inline microservices/payment-stripe/src/entities/coupon.ts]:
```suggestion
  private static isAmountOffExists(entity: Coupon): boolean {
```

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 49]
```diff
@@ -0,0 +1,97 @@
...
+import Coupon from '@entities/coupon';
+import TCurrency from '@interfaces/currency';
+import Factory from '@services/payment-gateway/factory';
+
+class CreateCouponInput {
+  @IsUndefinable()
+  @Length(1, 36)
+  userId?: string;
+
+  @IsUndefinable()
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/coupon/create.ts]: Prefferable order
```suggestion
@TypeDecodator()
  @IsUndefinable()
```

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 50]
```diff
@@ -0,0 +1,97 @@
...
+
+  @IsValidate(CreateCouponInput, (entity) => CreateCouponInput.isDurationRepeating(entity))
+  @IsNumber()
+  durationInMonths?: number;
+
+  @IsUndefinable()
+  @IsNumber()
+  maxRedemptions?: number;
+
+  @IsString({ each: true })
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/coupon/create.ts]: optiopnal params should be at he end

File: microservices/payment-stripe/src/entities/coupon.ts (hunk 42 above)
Comment (a-sava) [reply microservices/payment-stripe/src/entities/coupon.ts:31]: Other entities, such as product or price use id with entity name

File: microservices/payment-stripe/src/methods/coupon/create.ts [hunk 51]
```diff
@@ -0,0 +1,97 @@
...
+  products: string[];
+
+  /**
+   * Check if duration is repeating
+   */
+  private static isDurationRepeating(entity: CreateCouponInput) {
```
Comment (OlegDO) [inline microservices/payment-stripe/src/methods/coupon/create.ts]:
```suggestion
  private static isDurationRepeating(entity: CreateCouponInput): boolean {
```

File: microservices/payment-stripe/src/methods/coupon/create.ts (hunk 44 above)
Comment (a-sava) [reply microservices/payment-stripe/src/methods/coupon/create.ts:78]: i've changed it after that comment https://github.com/Lomray-Software/microservices/pull/69#discussion_r1288120850

File: microservices/payment-stripe/src/services/payment-gateway/abstract.ts [hunk 52]
```diff
@@ -225,6 +244,32 @@ abstract class Abstract {
...
+        message: "One or more products don't exists.",
+      });
+    }
+
+    const coupon = this.couponRepository.create({
```
Comment (OlegDO) [inline microservices/payment-stripe/src/services/payment-gateway/abstract.ts:265]: Declaration standard
image

**Review by OlegDO — CHANGES_REQUESTED** [review]

**Review by OlegDO — APPROVED** [review]

**Review by MatthewPattell — APPROVED** [review]

**Pull request merged**